dealiiadapter.cc

```diff
@@ -770,6 +791,124 @@ void ElasticProblem<dim>::output_results(const unsigned int timestep) const
+void ElasticProblem<dim>::initialize_precice()
+{
+    //get n_interface_nodes
+    std::set<types::boundary_id> couplingBoundary;
+    couplingBoundary.insert(parameters.interface_mesh_ID);
+    const FEValuesExtractors::Scalar x_displacement(0);
+
+    DoFTools::extract_boundary_dofs(dof_handler, fe.component_mask(x_displacement), coupling_dofs, couplingBoundary);
+    n_interface_nodes = coupling_dofs.n_elements();//TODO: Adapt
+
+    write_data = new double [dim * n_interface_nodes];
+    read_data  = new double [dim * n_interface_nodes];
+    interface_nodes_positions = new double [dim*n_interface_nodes];
+    interface_nodes_IDs       = new int [n_interface_nodes];
+
+    // get nodes_positions
+    std::map<types::global_dof_index, Point<dim>> support_points;
+    DoFTools::map_dofs_to_support_points(MappingQ1<dim>(), dof_handler, support_points);
+    int position_iterator=0;
+    for (IndexSet::ElementIterator element=coupling_dofs.begin(); element!=coupling_dofs.end(); element++)
+    {
+        for(int jj=0; jj<dim; jj++)
+            interface_nodes_positions[position_iterator * dim + jj] = support_points[*element][jj];
+
+        position_iterator++;
+    }
+
+    // store initial write_data
+    get_write_data(write_data);
+
+    precice.setMeshVertices(meshID, n_interface_nodes, interface_nodes_positions, interface_nodes_IDs);
+
+    precice.initialize();
+
+    //Write initial writeData to preCICE
+    if (precice.isActionRequired(precice::constants::actionWriteInitialData()))
+    {
+        precice.writeBlockVectorData(write_data_ID, n_interface_nodes, interface_nodes_IDs, write_data);
+        precice.fulfilledAction(precice::constants::actionWriteInitialData());
+    }
+
+    precice.initializeData();
+
+    //Read initial readData from preCICE for the first time step
+    if (precice.isReadDataAvailable())
+        precice.readBlockVectorData(read_data_ID, n_interface_nodes, interface_nodes_IDs, read_data);
+
+    get_read_data(read_data);
+
+}
+
+template <int dim>
+void ElasticProblem<dim>::exchange_precice_data()
+{
+    if(precice.isWriteDataRequired(time.get_delta_t()))
+    {
+        get_write_data(write_data);
+        precice.writeBlockVectorData(write_data_ID, n_interface_nodes, interface_nodes_IDs, write_data);
+    }
+
+    precice.advance(time.get_delta_t());
+
+    if(precice.isReadDataAvailable())
+    {
+        precice.readBlockVectorData(read_data_ID, n_interface_nodes, interface_nodes_IDs, read_data);
+        get_read_data(read_data);
+    }
+}
+
+template <int dim>
+void ElasticProblem<dim>::finalize_precice()
+{
+    delete[] read_data;
+    delete[] write_data;
+    delete[] interface_nodes_IDs;
+    delete[] interface_nodes_positions;
+
+    precice.finalize();
+}
+
+template <int dim>
+void ElasticProblem<dim>::get_write_data(double *write_data)
+{
+    int data_iterator = 0;
+    for (IndexSet::ElementIterator element=coupling_dofs.begin(); element!=coupling_dofs.end(); element++)
```

Comment:
Sad that IndexSet does not satisfy the requirements to use a range for loop. You can use `auto` though.
```suggestion
    for (auto element=coupling_dofs.begin(); element!=coupling_dofs.end(); ++element)
```